File: Table/symbol_table.py
# x = 5
# name : x
# type : int
# value : 5
# lod = 1
# scope = global
# role = variable
# size = 4

class Symbol:
    def __init__(self, name, type, value, line, scope, role, size):
        self.name = name
        self.type = type
        self.value = value
        self.line = line
        self.scope = scope
        self.role = role
        self.size = size

    def __repr__(self):
        return f"Symbol({self.name}, {self.type}, {self.value}, {self.line}, {self.scope}, {self.role}, {self.size})"
    
class SymbolTable:
    def __init__(self):
        self.scope_stack = [{}]

    def current_scope(self):
        return self.scope_stack[-1]
    
    def insert(self, name, type, value, line, scope, role, size):
        current = self.current_scope()
        if name in current:
            raise Exception(f"Symbol {name} already exists in the current scope")
        symbol = Symbol(name=name, type=type, value=value, line=line, scope=scope, role=role, size=size)
        current[name] = symbol # Adding X to the 2nd level/scope

    def lookup(self, name):
        for scope in reversed(self.scope_stack):
            if name in scope:
                return scope[name]
        raise Exception(f"Symbol {name} not found in any scope")
    
    def update(self, name, value, new_value):
        symbol = self.lookup(name)
        if symbol:
            symbol.value = new_value
        else:
            raise Exception(f"Symbol {name} not found in any scope")
        
    def enter_scope(self):
        self.scope_stack.append({})
        print('Entering new scope')

    def exit_scope(self):
        if len(self.scope_stack) > 1:
            self.scope_stack.pop()
            print('Exiting scope')
        else:
            raise Exception("Cannot exit global scope")
        
    def print_table(self):
        for i, scope in enumerate(self.scope_stack):
            print(f"Scope {i}:")
            for name, symbol in scope.items():
                print(f" {name}")
                print()
                
                #remake in java for discussion

File: Table/symbolTable.java
// x = 5
// name : x
// type : int
// value : 5
// lod = 1
// scope = global
// role = variable
// size = 4

import java.util.*;

class Symbol {
	String name;
	String type;
	Object value;
	int line;
	String scope;
	String role;
	int size;

	public Symbol(String name, String type, Object value, int line, String scope, String role, int size) {
		this.name = name;
		this.type = type;
		this.value = value;
		this.line = line;
		this.scope = scope;
		this.role = role;
		this.size = size;
	}

	@Override
	public String toString() {
		return "Symbol(" + name + ", " + type + ", " + value + ", " + line + ", " + scope + ", " + role + ", " + size
				+ ", ";
	}
}

class symbolTable {
	private Stack<Map<String, Symbol>> scopeStack;

	public void SymbolTable() {
		scopeStack = new Stack<>();
		scopeStack.push(new HashMap<>());
	}

	private Map<String, Symbol> currentScope() {
		return scopeStack.peek();
	}

	public void insert(String name, String type, Object value, int line, String scope, String role, int size) {
		Map<String, Symbol> current = currentScope();
		if (current.containsKey(name)) {
			throw new RuntimeException("Symbol " + name + "already exists in the current scope.");
		}

		Symbol symbol = new Symbol(name, type, value, line, scope, role, size);
		current.put(name, symbol);
	}

	public Symbol lookup (String name) {
		for (int i = scopeStack.size() - 1; i >= 0; i--) {
			Map <String, Symbol> scope = scopeStack.get(i);
			if (scope.containsKey(name)) {
				return scope.get(name);
			}
		}
		throw new RuntimeException("Symbol " + name + " not found in any scope.");
	}

	public void update(String name, Object newValue) {
		Symbol symbol = lookup(name);
		Symbol.value = newValue;
	}

	public void enterScope() {
		scopeStack.push(new HashMap<>());
		System.out.printIn("Entering new scope");
	}
	
}
